range 函数在 Python 中用于生成一个整数序列。该函数有三个参数：

start：序列的起始值，默认为 0。
stop：序列的结束值，但不包含此值。
step：序列中每个数的增量，默认为 1。
在你提供的代码 range(10, 1, -2) 中：

10 是序列的起始值。
1 是序列的结束值，但由于 range 不包含结束值，所以序列中最大的数会小于 1。
-2 是步长，表示序列中每个数之间的差值。

In [1]:
for i in range(10, 1, -2):
    print(i)

10
8
6
4
2


In [2]:
for i in 'python':
    if i == 'h':
        break
    print(i)

p
y
t


In [3]:
str1 = ''
str2 = ' '
if not str1:
    print(1)
elif not str2:
    print(2)
else:
    print(0)

1


我们有一个字符串 str1 被初始化为空字符串 ('')，另一个字符串 str2 被初始化为包含一个空格的字符串 (' ')。

让我们逐行分析代码：

str1 = ''：str1 被设置为空字符串。
str2 = ' '：str2 被设置为包含一个空格的字符串。
if not str1:：这个条件检查 str1 是否为假值。在Python中，空字符串 ('') 被认为是假值（False），所以这个条件为真。
print(1)：由于 if 条件为真，执行 print(1)。
elif not str2:：这个条件检查 str2 是否为假值。由于 str2 是一个包含空格的字符串，它在Python中被认为是真值（True），所以这个条件为假，不会执行。
else:：由于 if 条件为真，else 块不会被执行。

In [4]:
str = "Hello,Python";
suffix = "Python";
print (str.endswith(suffix,2));

True


str.endswith(suffix, 2) 会从字符串 str 的索引 2（从 0 开始计数）处开始检查是否以 suffix 结尾。

In [5]:
strs = 'abcd'
print(strs.center(10, '*'))

***abcd***


代码使用了字符串的 center 方法，其作用是将原字符串 strs 居中放置，并在两侧填充指定的字符，直到结果字符串达到指定的宽度

In [6]:
dicts = {'one': 1, 'two': 2, 'three': 3}
tmp = dicts.copy()
tmp['one'] = 'abc'
print(dicts)
print(tmp)

{'one': 1, 'two': 2, 'three': 3}
{'one': 'abc', 'two': 2, 'three': 3}


In [7]:
res = 0
for i in range(1, 4):
    for j in range(1, 4):
        for k in range(1, 4):
            if i != j and i != k and j != k:
                res += 1
print(res)

6


In [8]:
import copy
a = [1, 2, 3, 4, ['a', 'b']]
b = a
c = copy.copy(a)
d = copy.deepcopy(a)
a.append(5)
a[4].append('c')
print(a)
print(b)
print(c)
print(d)

[1, 2, 3, 4, ['a', 'b', 'c'], 5]
[1, 2, 3, 4, ['a', 'b', 'c'], 5]
[1, 2, 3, 4, ['a', 'b', 'c']]
[1, 2, 3, 4, ['a', 'b']]


浅拷贝（Shallow Copy）：浅拷贝会创建一个新对象，其内容是原对象中对象的引用拷贝。对于简单数据类型（如整数、浮点数、字符串等），这没有问题，因为它们是不可变的。但对于可变数据类型（如列表、字典等），浅拷贝会导致原始对象和拷贝对象共享相同的子对象。

如果您修改了原始对象 a 中的可变子对象（例如，修改 a 中的列表），那么这些修改也会反映在浅拷贝对象 c 中，因为 c 只是复制了 a 中子对象的引用。

同样，如果您修改了浅拷贝对象 c 中的可变子对象，这些修改也会反映在原始对象 a 中，因为它们共享相同的子对象。

深拷贝（Deep Copy）：深拷贝会创建一个新对象，并且递归地复制对象中的所有子对象。这意味着原始对象和拷贝对象之间没有任何共享的子对象。

如果您修改了原始对象 a 中的可变子对象，这些修改不会反映在深拷贝对象 d 中，因为 d 中的子对象是独立的拷贝。

同样，如果您修改了深拷贝对象 d 中的可变子对象，这些修改也不会反映在原始对象 a 中。

In [9]:
a = [['1']*3 for i in range(3)]

print(a)

[['1', '1', '1'], ['1', '1', '1'], ['1', '1', '1']]


range(3) 生成一个从 0 到 2 的序列。
对于序列中的每个 i（即 0、1、2），执行 ['1']*3 表达式，这会创建一个包含 3 个字符串 '1' 的列表。
这些列表被收集起来，构成一个新的列表 a。

In [10]:
b = [['1']]*3

print(b)

[['1'], ['1'], ['1']]


In [13]:
c=[]

for i in range(3):

    lis = ['1']*3

    c.append(lis)

print(c)

[['1', '1', '1'], ['1', '1', '1'], ['1', '1', '1']]


In [12]:
d = []

lis = ['1']*3

for i in range(3):

    d.append(lis)

print(d)

[['1', '1', '1'], ['1', '1', '1'], ['1', '1', '1']]


In [14]:
a = 10
b = 3
print(a // b)
print(a % b)
print(a / b)

3
1
3.3333333333333335


a // b：这是整数除法（也称为地板除法），它返回两个数相除后的整数部分。在这个例子中，10 // 3 的结果是 3，因为 10 除以 3 等于 3 余 1，整数部分是 3。

a % b：这是取模运算，它返回两个数相除后的余数。在这个例子中，10 % 3 的结果是 1，因为 10 除以 3 等于 3 余 1，余数是 1。

a / b：这是浮点数除法，它返回两个数相除后的精确结果，包括小数部分。在这个例子中，10 / 3 的结果是 3.3333333333333335，因为 10 除以 3 等于 3.3333333333333335。

In [15]:
class Base(object):

    count = 0

    def __init__(self):

        pass
b1 = Base()
b2 = Base()

b1.count = b1.count + 1

print(b1.count,end=" ")

print(Base.count,end=" ")

print(b2.count)

1 0 0


In [ ]:
for i in range(5):
    if i == 2:
        pass
    print(i) # pass 语句是一个空操作，它表示“什么都不做”

0
1
2
3
4


In [17]:
num = 1

def fn():

    num += 1

    return lambda:print(num)

x = fn()

x()

UnboundLocalError: cannot access local variable 'num' where it is not associated with a value